### ISSUE TITLE
Memory Error while loading Kinetic Dataset

### ISSUE THREAD

**Issue by hungdtrn** [facebookresearch/SlowFast#551]
Hi, I am trying to reproduce the results of the MViT model on the Kinetic dataset.
However, I encounter this error while loading the Kinectic dataset

`[06/01 12:11:45][INFO] kinetics.py:   88: Constructing Kinetics test...
[06/01 12:11:46][INFO] kinetics.py:  152: Constructing kinetics dataloader (size: 386710 skip_rows 0) from ../tmp_slowfast/cv2_data/test.csv 
[06/01 12:11:46][INFO] test_net.py:  211: Testing model for 6043 iterations
double free or corruption (!prev)
malloc(): invalid size (unsorted)
double free or corruption (!prev)
Traceback (most recent call last):
  File "/home/tduy/anaconda3/envs/slowfast/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 986, in _try_get_data
    data = self._data_queue.get(timeout=timeout)
  File "/home/tduy/anaconda3/envs/slowfast/lib/python3.8/queue.py", line 179, in get
    self.not_empty.wait(remaining)
  File "/home/tduy/anaconda3/envs/slowfast/lib/python3.8/threading.py", line 306, in wait
    gotit = waiter.acquire(True, timeout)
  File "/home/tduy/anaconda3/envs/slowfast/lib/python3.8/site-packages/torch/utils/data/_utils/signal_handling.py", line 66, in handler
    _error_if_any_worker_fails()
RuntimeError: DataLoader worker (pid 4035577) is killed by signal: Aborted. 

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "tools/run_net.py", line 45, in <module>
    main()
  File "tools/run_net.py", line 30, in main
    launch_job(cfg=cfg, init_method=args.init_method, func=test)
  File "/home/tduy/Workspace/git/newest_slowfast/slowfast/utils/misc.py", line 311, in launch_job
    func(cfg=cfg)
  File "/home/tduy/Workspace/git/newest_slowfast/tools/test_net.py", line 244, in test
    test_meter = perform_test(test_loader, model, test_meter, cfg, writer)
  File "/home/tduy/anaconda3/envs/slowfast/lib/python3.8/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/home/tduy/Workspace/git/newest_slowfast/tools/test_net.py", line 50, in perform_test
    for cur_iter, (inputs, labels, video_idx, time, meta) in enumerate(
  File "/home/tduy/anaconda3/envs/slowfast/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 517, in __next__
    data = self._next_data()
  File "/home/tduy/anaconda3/envs/slowfast/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 1182, in _next_data
    idx, data = self._get_data()
  File "/home/tduy/anaconda3/envs/slowfast/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 1138, in _get_data
    success, data = self._try_get_data()
  File "/home/tduy/anaconda3/envs/slowfast/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 999, in _try_get_data
    raise RuntimeError('DataLoader worker (pid(s) {}) exited unexpectedly'.format(pids_str)) from e
RuntimeError: DataLoader worker (pid(s) 4035577) exited unexpectedly
`

It seems that the error is related to the buffer allocation in the `decode` function inside the `kinetic.py` file. 
Does anybody know how to fix this?

**Comment by XinyuSun** [comment]
Hi! I meet the same problem. Any solution?

**Comment by haooooooqi** [comment]
Maybe it is some data issue? I don't currently seeing this issue...

**Comment by guochengqian** [comment]
looks like an OOM issue. Try to use a lower batch size might work.

**Comment by guochengqian** [comment]
same problem here. 
I found the issue is from: `torch.ops.video_reader.read_video_from_memory`
add environment: `MALLOC_CHECK_=2`, the error changes to: `free(): invalid pointer`

Any clue how to fix this?


@haooooooqi  Could you provide an updated instruction how to install the correct environment. Now, I find many people are struggling using slowfast codebase.

**Comment by hungdtrn** [comment]
> same problem here.
> 
>     1. I found the issue is from: `torch.ops.video_reader.read_video_from_memory`
> 
>     2. add environment: `MALLOC_CHECK_=2`, the error changes to: `free(): invalid pointer`
> 
>     3. Using pyav backend, the issue can be avoided.
> 
> 
> Any clue how to make this library runable using torchvision backend?
> 
> @haooooooqi Could you provide an updated instruction how to install the correct environment. Now, I find many people are struggling using slowfast codebase.
> 
> My environment: (python 3.8, nightly versions for pytorch, torchvision (from source).
> 
> ```
> name: slowfast
> channels:
>   - pytorch-nightly
>   - conda-forge
>   - defaults
> dependencies:
>   - _libgcc_mutex=0.1=conda_forge
>   - _openmp_mutex=4.5=2_kmp_llvm
>   - aom=3.4.0=h27087fc_1
>   - av=9.2.0=py38hb0ab478_2
>   - blas=1.0=mkl
>   - bzip2=1.0.8=h7f98852_4
>   - ca-certificates=2022.6.15=ha878542_0
>   - cudatoolkit=11.3.1=h2bc3f7f_2
>   - expat=2.4.8=h27087fc_0
>   - ffmpeg=4.4.2=gpl_hfe78399_107
>   - font-ttf-dejavu-sans-mono=2.37=hab24e00_0
>   - font-ttf-inconsolata=3.000=h77eed37_0
>   - font-ttf-source-code-pro=2.038=h77eed37_0
>   - font-ttf-ubuntu=0.83=hab24e00_0
>   - fontconfig=2.14.0=h8e229c2_0
>   - fonts-conda-ecosystem=1=0
>   - fonts-conda-forge=1=0
>   - freetype=2.12.1=hca18f0e_0
>   - gettext=0.19.8.1=h73d1719_1008
>   - gmp=6.2.1=h58526e2_0
>   - gnutls=3.7.7=hf3e180e_0
>   - icu=70.1=h27087fc_0
>   - jpeg=9e=h166bdaf_2
>   - lame=3.100=h7f98852_1001
>   - lcms2=2.12=hddcbb42_0
>   - ld_impl_linux-64=2.36.1=hea4e1c9_2
>   - lerc=4.0.0=h27087fc_0
>   - libdeflate=1.13=h166bdaf_0
>   - libdrm=2.4.113=h166bdaf_0
>   - libffi=3.4.2=h7f98852_5
>   - libgcc-ng=12.1.0=h8d9b700_16
>   - libiconv=1.16=h516909a_0
>   - libidn2=2.3.3=h166bdaf_0
>   - libnsl=2.0.0=h7f98852_0
>   - libpciaccess=0.16=h516909a_0
>   - libpng=1.6.37=h753d276_4
>   - libsqlite=3.39.2=h753d276_1
>   - libstdcxx-ng=12.1.0=ha89aaad_16
>   - libtasn1=4.19.0=h166bdaf_0
>   - libtiff=4.4.0=h0e0dad5_3
>   - libunistring=0.9.10=h7f98852_0
>   - libuuid=2.32.1=h7f98852_1000
>   - libva=2.15.0=h166bdaf_0
>   - libvpx=1.11.0=h9c3ff4c_3
>   - libwebp-base=1.2.4=h166bdaf_0
>   - libxcb=1.13=h7f98852_1004
>   - libxml2=2.9.14=h22db469_4
>   - libzlib=1.2.12=h166bdaf_2
>   - llvm-openmp=14.0.4=he0ac6c6_0
>   - mkl=2021.4.0=h8d4b97c_729
>   - mkl-service=2.4.0=py38h95df7f1_0
>   - mkl_fft=1.3.1=py38h8666266_1
>   - mkl_random=1.2.2=py38h1abd341_0
>   - ncurses=6.3=h27087fc_1
>   - nettle=3.8.1=hc379101_1
>   - ninja=1.11.0=h924138e_0
>   - numpy=1.23.1=py38h6c91a56_0
>   - numpy-base=1.23.1=py38ha15fc14_0
>   - openh264=2.3.0=h27087fc_0
>   - openjpeg=2.5.0=h7d73246_1
>   - openssl=3.0.5=h166bdaf_1
>   - p11-kit=0.24.1=hc5aa10d_0
>   - pillow=9.2.0=py38ha3b2c9c_2
>   - pip=22.2.2=pyhd8ed1ab_0
>   - pthread-stubs=0.4=h36c2ea0_1001
>   - python=3.8.13=ha86cf86_0_cpython
>   - python_abi=3.8=2_cp38
>   - pytorch=1.13.0.dev20220904=py3.8_cuda11.3_cudnn8.3.2_0
>   - pytorch-mutex=1.0=cuda
>   - readline=8.1.2=h0f457ee_0
>   - setuptools=65.3.0=pyhd8ed1ab_1
>   - six=1.16.0=pyh6c4a22f_0
>   - sqlite=3.39.2=h4ff8645_1
>   - svt-av1=1.2.1=h27087fc_0
>   - tbb=2021.5.0=h924138e_1
>   - tk=8.6.12=h27826a3_0
>   - typing_extensions=4.3.0=pyha770c72_0
>   - wheel=0.37.1=pyhd8ed1ab_0
>   - x264=1!164.3095=h166bdaf_2
>   - x265=3.5=h924138e_3
>   - xorg-fixesproto=5.0=h7f98852_1002
>   - xorg-kbproto=1.0.7=h7f98852_1002
>   - xorg-libx11=1.7.2=h7f98852_0
>   - xorg-libxau=1.0.9=h7f98852_0
>   - xorg-libxdmcp=1.1.3=h7f98852_0
>   - xorg-libxext=1.3.4=h7f98852_1
>   - xorg-libxfixes=5.0.3=h7f98852_1004
>   - xorg-xextproto=7.3.0=h7f98852_1002
>   - xorg-xproto=7.0.31=h7f98852_1007
>   - xz=5.2.6=h166bdaf_0
>   - zstd=1.5.2=h6239696_4
>   - pip:
>     - absl-py==1.2.0
>     - antlr4-python3-runtime==4.9.3
>     - black==22.3.0
>     - cachetools==5.2.0
>     - certifi==2022.6.15
>     - charset-normalizer==2.1.1
>     - click==8.1.3
>     - cloudpickle==2.1.0
>     - cycler==0.11.0
>     - detectron2==0.6
>     - fairscale==0.4.8
>     - fonttools==4.37.1
>     - future==0.18.2
>     - fvcore==0.1.5
>     - google-auth==2.11.0
>     - google-auth-oauthlib==0.4.6
>     - grpcio==1.48.1
>     - hydra-core==1.2.0
>     - idna==3.3
>     - importlib-metadata==4.12.0
>     - importlib-resources==5.9.0
>     - iopath==0.1.9
>     - joblib==1.1.0
>     - kiwisolver==1.4.4
>     - markdown==3.4.1
>     - markupsafe==2.1.1
>     - matplotlib==3.5.3
>     - multimethod==1.8
>     - mypy-extensions==0.4.3
>     - networkx==2.8.6
>     - oauthlib==3.2.0
>     - omegaconf==2.2.3
>     - opencv-python==4.6.0.66
>     - packaging==21.3
>     - pandas==1.4.4
>     - parameterized==0.8.1
>     - pathspec==0.10.1
>     - platformdirs==2.5.2
>     - portalocker==2.5.1
>     - protobuf==3.19.4
>     - psutil==5.9.2
>     - pyasn1==0.4.8
>     - pyasn1-modules==0.2.8
>     - pycocotools==2.0.4
>     - pydot==1.4.2
>     - pyparsing==3.0.9
>     - python-dateutil==2.8.2
>     - pytorchvideo==0.1.5
>     - pytz==2022.2.1
>     - pyyaml==6.0
>     - requests==2.28.1
>     - requests-oauthlib==1.3.1
>     - rsa==4.9
>     - scikit-learn==1.1.2
>     - scipy==1.9.1
>     - shortuuid==1.0.9
>     - simplejson==3.17.6
>     - tabulate==0.8.10
>     - tensorboard==2.10.0
>     - tensorboard-data-server==0.6.1
>     - tensorboard-plugin-wit==1.8.1
>     - termcolor==1.1.0
>     - threadpoolctl==3.1.0
>     - timm==0.6.7
>     - tomli==2.0.1
>     - torchvision==0.14.0a0+56fb841
>     - tqdm==4.64.1
>     - urllib3==1.26.12
>     - werkzeug==2.2.2
>     - yacs==0.1.8
>     - zipp==3.8.1
> prefix: /home/qiang/anaconda3/envs/slowfast
> ```

Hi, I've managed to fix the problem by installing ffmpeg version 4.2 `conda install ffmpeg=4.2`
This seems to be an error of torchvision.
For more information, you can visit here https://github.com/pytorch/vision/issues/5419

**Comment by guochengqian** [comment]
> 

wow! Thanks for the news. will try.

**Comment by khizar-anjum** [comment]
I recreated the same environment as @guochengqian posted above, downgraded the ffmpeg to 4.2 with torchvision backed (had to compile it myself for video-reader support) but it still gives me the double free or corruption error.

**Comment by khizar-anjum** [comment]
Alright, so after a long week spent on making this SlowFast library work (with `video_reader` backend), I have finally made it work irrespective of the `python` or `ffmpeg` version (although python 3.9 would not work in any case). 

### TLDR  
You can use this gist I posted: [kinetics_cleaner.ipynb](https://gist.github.com/khizar-anjum/260f966b27b8fd1eb19eb2b28199619c) and follow the steps in it. The most important part of the code is __Section 3__, which decodes each video file with you current configuration and gives you the files for which the decoding failed. These then you can remove from your `.csv` list.  

### Analysis of the problem
The problem is that while using the `video_reader` backend on Pytorch, the API crashes by either throwing `malloc (): double free or corruption` or some other segmentation fault. As far as I have seen, this problem is linked somehow to the combination of `ffmpeg` , `pytorch` and `torchvision` versions on your system. During my experiments, I have seen that when using `ffmpeg==4.2`, there would be a memory error, while the video will be perfectly decoded using `ffmpeg==5.1.2`. You can access the relevant discussions at [pytorch/vision#5419](https://github.com/pytorch/vision/issues/5419) and [pytorch/vision#6204](https://github.com/pytorch/vision/issues/6204).  
This can be avoid using `PyAV` backend but that one is extremely slow.  

###  How to go about fixing it?
Hence, as far as I reckon, whatever configuration you are using, the set of un-decodable videos can be different. So, a very simple method would be to then fix your packages configuration, and then just remove the videos which are un-decodable. This can be done by using a simple script (as mentioned in  [pytorch/vision#6204](https://github.com/pytorch/vision/issues/6204)) : 
```
from torchvision.io import read_video
from torchvision import set_video_backend
video, audio, info = read_video(VIDEO_NAME, pts_unit = 'sec')
```
The [gist](https://gist.github.com/khizar-anjum/260f966b27b8fd1eb19eb2b28199619c) I provided does exactly this in section 3, wherein it gives you the names of the videos which cannot be decoded under your configuration. 
I can provide the 928 files I removed from my dataset on request but not doing so as I think many of them would still be decodable under other configurations, and vice versa.  

I hope this saves somebody time! Cheers!

P.S. You will have to recompile `torchvision` from scratch if you want to use `video_reader` backend. My configuration currently is `ffmpeg==5.1.2`, `torchvision==0.14.0`, `pytorch==1.13.0`, `cuda==11.1` and `gcc=10.4`. The OS is `Ubuntu 22.04`. Follow @guochengqian's comment for other packages' versions.

**Comment by Saba003** [comment]
Hello,
I am training slowfast network using kinetics data format,My configuration look like following 

Data Format

``
data
└── Kinetics
    ├── ClassA
    │   └── cooking_egg.mp4
    ├── ClassB
    │   └── feeding_birds.mp4
    ├── ClassC
    │   └── knitting.mp4
    ├── ClassD
    │   └── playing_chess.mp4
    ├── ClassE
    │   └── playing_guitar.mp4
    ├── classids.json
    ├── test.csv
    ├── train.csv
    └── val.csv
``

Configuration File
I am using SLOWFAST_8x8_R50.yaml file

```

TRAIN:
  ENABLE: True
  DATASET: Kinetics
  BATCH_SIZE: 1
  EVAL_PERIOD: 10
  CHECKPOINT_PERIOD: 1
  AUTO_RESUME: True
  CHECKPOINT_FILE_PATH: /media/Anomaly_Research/SlowFast/SlowFast/slowfast/models/SLOWFAST_8x8_R50.pkl
  CHECKPOINT_TYPE: caffe2
DATA:
  NUM_FRAMES: 32
  SAMPLING_RATE: 2
  TRAIN_JITTER_SCALES: [256, 320]
  TRAIN_CROP_SIZE: 224
  TEST_CROP_SIZE: 256
  INPUT_CHANNEL_NUM: [3, 3]
  PATH_TO_DATA_DIR: /media/Anomaly_Research/SlowFast/SlowFast/data/Kinetics
  DECODING_BACKEND: torchvision
SLOWFAST:
  ALPHA: 4
  BETA_INV: 8
  FUSION_CONV_CHANNEL_RATIO: 2
  FUSION_KERNEL_SZ: 7
RESNET:
  ZERO_INIT_FINAL_BN: True
  WIDTH_PER_GROUP: 64
  NUM_GROUPS: 1
  DEPTH: 50
  TRANS_FUNC: bottleneck_transform
  STRIDE_1X1: False
  NUM_BLOCK_TEMP_KERNEL: [[3, 3], [4, 4], [6, 6], [3, 3]]
  SPATIAL_STRIDES: [[1, 1], [2, 2], [2, 2], [2, 2]]
  SPATIAL_DILATIONS: [[1, 1], [1, 1], [1, 1], [1, 1]]
NONLOCAL:
  LOCATION: [[[], []], [[], []], [[], []], [[], []]]
  GROUP: [[1, 1], [1, 1], [1, 1], [1, 1]]
  INSTANTIATION: dot_product
BN:
  USE_PRECISE_STATS: True
  NUM_BATCHES_PRECISE: 200
SOLVER:
  BASE_LR: 0.1
  LR_POLICY: cosine
  MAX_EPOCH: 196
  MOMENTUM: 0.9
  WEIGHT_DECAY: 1e-4
  WARMUP_EPOCHS: 34.0
  WARMUP_START_LR: 0.01
  OPTIMIZING_METHOD: sgd
MODEL:
  NUM_CLASSES: 6
  ARCH: slowfast
  MODEL_NAME: SlowFast
  LOSS_FUNC: cross_entropy
  DROPOUT_RATE: 0.5
TEST:
  ENABLE: False
  DATASET: kinetics
  BATCH_SIZE: 1
DATA_LOADER:
  NUM_WORKERS: 8
  PIN_MEMORY: True
NUM_GPUS: 0
NUM_SHARDS: 1
RNG_SEED: 0
OUTPUT_DIR: ./demo

DEMO:
  ENABLE: False
  LABEL_FILE_PATH: data/MyData/classids.json # Path to json file providing class_name - id mapping.
  INPUT_VIDEO: data/MyData/train.csv# Path to input video file.
  OUTPUT_FILE: data/MyData # Path to output video file to write results to.
               # Leave an empty string if you would like to display results to a window.
  THREAD_ENABLE: # Run video reader/writer in the background with multi-threading.
  NUM_VIS_INSTANCES: # Number of CPU(s)/processes use to run video visualizer.
  NUM_CLIPS_SKIP: # Number of clips to skip prediction/visualization
                  # (mostly to smoothen/improve display quality with wecam input).


```


Failed to decode by torchvision with exception: '_OpNamespace' 'video_reader' object has no attribute 'probe_video_from_memory'
[06/17 01:46:21][WARNING] kinetics.py:  384: Failed to decode video idx 2 from /media/Anomaly_Research/SlowFast/SlowFast/data/Kinetics/ClassC/knitting.mp4; trial 97
Failed to decode by torchvision with exception: '_OpNamespace' 'video_reader' object has no attribute 'probe_video_from_memory'
[06/17 01:46:21][WARNING] kinetics.py:  384: Failed to decode video idx 2 from /media/Anomaly_Research/SlowFast/SlowFast/data/Kinetics/ClassC/knitting.mp4; trial 98
Failed to decode by torchvision with exception: '_OpNamespace' 'video_reader' object has no attribute 'probe_video_from_memory'
[06/17 01:46:21][WARNING] kinetics.py:  384: Failed to decode video idx 2 from /media/Anomaly_Research/SlowFast/SlowFast/data/Kinetics/ClassC/knitting.mp4; trial 99
[06/17 01:46:21][WARNING] kinetics.py:  509: Failed to fetch video after 100 retries.
Traceback (most recent call last):
  File "/media/Anomaly_Research/SlowFast/SlowFast/tools/run_net.py", line 52, in <module>
    main()
  File "/media/Anomaly_Research/SlowFast/SlowFast/tools/run_net.py", line 27, in main
    launch_job(cfg=cfg, init_method=args.init_method, func=train)
  File "/home/anaconda3/envs/slowFast/lib/python3.11/site-packages/slowfast/utils/misc.py", line 430, in launch_job
    func(cfg=cfg)
  File "/media/mypc/Anomaly_Research/SlowFast/SlowFast/tools/train_net.py", line 682, in train
    train_epoch(
  File "/media/Anomaly_Research/SlowFast/SlowFast/tools/train_net.py", line 77, in train_epoch
    for cur_iter, (inputs, labels, index, time, meta) in enumerate(
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: cannot unpack non-iterable NoneType object

I am facing the above error.Can you help me in this regard.thanks

**Comment by khizar-anjum** [comment]
@Saba003 You need to build torchvision from scratch with video_reader backend support.

**Comment by gxxo** [comment]
@Saba003  did you solve this "TypeError: cannot unpack non-iterable NoneType object" problem?
